liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy

```diff
@@ -0,0 +1,272 @@
+            @Override
+            ValidationErrors validate(Database database) {
+                ValidationErrors changeValidationErrors = new ValidationErrors()
+                changeValidationErrors.addError("Test message")
+                return changeValidationErrors;
+            }
+        })
+
+        List<RanChangeSet> ran = new ArrayList<RanChangeSet>()
+        ran.add(new RanChangeSet(changeSet))
+        ValidatingVisitor handler = new ValidatingVisitor(ran)
+        handler.visit(changeSet, new DatabaseChangeLog(), null, null)
+
+        then:
+        handler.getSetupExceptions().size() == 0
+        handler.validationPassed()
+    }
+
+    void "validate successful visit with single valid dbms set"() throws Exception {
+        when:
+        CreateTableChange change = new CreateTableChange()
+        change.setTableName("table1")
+        ColumnConfig column1 = new ColumnConfig()
+        change.addColumn(column1)
+        column1.setName("col1")
+        column1.setType("int")
+
+        ChangeSet changeSet = new ChangeSet("1", "testAuthor", false, false, "path/changelog", null, "postgresql", null)
+        changeSet.addChange(change)
+        ValidatingVisitor handler = new ValidatingVisitor(new ArrayList<RanChangeSet>())
+        handler.visit(changeSet, new DatabaseChangeLog(), new MockDatabase(), null)
+
+        then:
+        handler.validationPassed()
+    }
+
+    void "validate successful visit with valid list of dbms set"() throws Exception {
+        when:
+        CreateTableChange change = new CreateTableChange()
+        change.setTableName("table1")
+        ColumnConfig column1 = new ColumnConfig()
+        change.addColumn(column1)
+        column1.setName("col1")
+        column1.setType("int")
+        ChangeSet changeSet = new ChangeSet("1", "testAuthor", false, false, "path/changelog", null, "postgresql, mssql, h2", null)
+        changeSet.addChange(change)
+
+        ValidatingVisitor handler = new ValidatingVisitor(new ArrayList<RanChangeSet>())
+        handler.visit(changeSet, new DatabaseChangeLog(), new MockDatabase(), null)
+
+        then:
+        handler.validationPassed()
+    }
+
+    void "validate unsuccessful visit with invalid dbms set"() throws Exception {
+        when:
+        CreateTableChange change = new CreateTableChange()
+        change.setTableName("table1")
+        ColumnConfig column1 = new ColumnConfig()
+        change.addColumn(column1)
+        column1.setName("col1")
+        column1.setType("int")
+        ChangeSet changeSet = new ChangeSet("1", "testAuthor", false, false, "path/changelog", null, "post", null)
+        changeSet.addChange(change)
+
+        ValidatingVisitor handler = new ValidatingVisitor(new ArrayList<RanChangeSet>())
+        handler.visit(changeSet, new DatabaseChangeLog(), new MockDatabase(), null)
+
+        then:
+        !handler.validationPassed()
+    }
+
+    void "validate successful visit with none dbms set"() throws Exception {
```

Comment:
The literal strings "none" and "all" are valid DBMS values? The purpose of "all" seems relatively straightforward, if not a bit redundant as the default is "all" when no dbms attribute is configured. However, what does it mean to have a changeset that has `dbms="none"`?

Reply:
Hey Erz, yeap they are. At least this is what the documentation states for this:

> Specifies which database type(s) a changeset is to be used for. See valid database type names on [Supported Databases](https://www.liquibase.org/get-started/databases). Separate multiple databases with commas. Specify that a changeset is not applicable to a particular database type by prefixing with !. The keywords all and none are also available.

Also, these are existent tests that I just migrated from JUnit to Groovy after adding some new tests related to this PR.